### ISSUE TITLE
[resolved] errors/timeouts when using reverse proxy

### ISSUE THREAD

**Issue by justsomeguyontheinternet** [minio/minio#10086]
hello. i'm having an issue with the latest release. in the web ui, after login it says "minio server is unreachable" when trying to list the contents of a bucket. chrome shows webrpc failed with message "err_incomplete_chunked_encoding". it appears to be some sort of timeout in minio. using an application built on the minio client errors are produced about downloaded files being smaller than expected, also suggesting some sort of timeout. the issue is not present when accessing minio over LAN, it only happens over WAN.

the issue is NOT present on RELEASE.2020-06-22T03-12-50Z

see slack discussion https://minio.slack.com/archives/C3NDUB8UA/p1595121370065500

## Expected Behavior
file get operations succeed without error
web ui lists bucket contents without error
mc ls listing bucket contents succeeds without error

## Current Behavior
web ui displays error "minio server is unreachable" when listing bucket contents. chrome shows webrpc failed with message "err_incomplete_chunked_encoding"
application built on minio client gives errors when downloading files something like "expected 50 MB file, got 23 MB"
mc ls gives error "<ERROR> Unable to list folder. unexpected EOF"
reverse proxy shows error "reverseproxy1   | 2020/07/19 01:10:39 [error] 22#22: *1 recv() failed (104: Connection reset by peer) while sending to client, client: ****, server: ****, request: "POST /minio/webrpc HTTP/1.1", upstream: "http://****:9000/minio/webrpc", host: "****:9000", referrer: "https://****:9000/minio/****/"

## Possible Solution
aggressive timeout setting? (see slack chat https://minio.slack.com/archives/C3NDUB8UA/p1595121370065500)

## Steps to Reproduce (for bugs)
create docker service using docker compose for nginx with reverse proxy (latest) following instructions in minio cookbook for nginx.conf
access webui, login, try to list bucket contents
OR using mc ls bucket contents

## Context
minio is unusable in this state. i want to run minio with reverse proxy as the reverse proxy has more TLS version flexibility then minio (needed to support older and/or diverse clients)

## Regression
i don't think so.

## Your Environment
ubuntu 19.10, minio latest, nginx latest as reverse proxy with tls encrypted connection
single local hard drive, single instance (not distributed)

**Comment by dpgarrick** [comment]
Looks like you're onto a fix but FWIW I'm getting a similar issue using caddy as a reverse proxy and I'm sure it must be related. I have been using Caddy 2.1.1 for a couple weeks as a reverse proxy with basic settings and had no issues until I updated to the latest minio. My caddyfile is just the basic reverse_proxy:

```
mydomain.com {
reverse_proxy 127.0.0.1:9002 {
  }
}
```

Previously using minio RELEASE.2020-06-22T03-12-50Z everything was running smoothly. On latest release I can no longer upload or download large files (>~20MB) via WAN either using the latest mc version or with a pre-signed URL. I tested from multiple client machines around the world and they all now repeatedly get timeouts and errors of the form:

```
mc: <ERROR> Failed to copy https://mydomain.com/bucket/a-1gb-file.gz. Input reader closed pre-maturely. Expected 1273465279 bytes, but only received 6421945 bytes.
```

I rolled minio back to RELEASE.2020-06-22T03-12-50Z and verified that I can again upload and download multiple large files from the minio server onto multiple machines with no issues.  

I also tried adding 

```
   transport http {
   keepalive 1d
   }
```

to the caddyfile as a test on latest minio but it did not help. 

`mc admin trace` shows no errors while the caddy logs just show `connection reset by peer`

**Comment by harshavardhana** [comment]
> Looks like you're onto a fix but FWIW I'm getting a similar issue using caddy as a reverse proxy and I'm sure it must be related. I have been using Caddy 2.1.1 for a couple weeks as a reverse proxy with basic settings and had no issues until I updated to the latest minio. My caddyfile is just the basic reverse_proxy:
> 
> ```
> mydomain.com {
> reverse_proxy 127.0.0.1:9002 {
>   }
> }
> ```
> 
> Previously using minio RELEASE.2020-06-22T03-12-50Z everything was running smoothly. On latest release I can no longer upload or download large files (>~20MB) via WAN either using the latest mc version or with a pre-signed URL. I tested from multiple client machines around the world and they all now repeatedly get timeouts and errors of the form:
> 
> ```
> mc: <ERROR> Failed to copy https://mydomain.com/bucket/a-1gb-file.gz. Input reader closed pre-maturely. Expected 1273465279 bytes, but only received 6421945 bytes.
> ```
> 
> I rolled minio back to RELEASE.2020-06-22T03-12-50Z and verified that I can again upload and download multiple large files from the minio server onto multiple machines with no issues.
> 
> I also tried adding
> 
> ```
>    transport http {
>    keepalive 1d
>    }
> ```
> 
> to the caddyfile as a test on latest minio but it did not help.
> 
> `mc admin trace` shows no errors while the caddy logs just show `connection reset by peer`

@justsomeguyontheinternet @dpgarrick please test the PR you may have to build a binary locally - use go1.13.13

**Comment by dpgarrick** [comment]
I'm running Ubuntu 18.04. Don't have time currently to build a local binary on go1.13.13 but following along in slack and this version you put on the play server DEVELOPMENT.2020-07-19T21-00-50Z seems to work on my initial testing @harshavardhana

**Comment by harshavardhana** [comment]
> I'm running Ubuntu 18.04. Don't have time currently to build a local binary on go1.13.13 but following along in slack and this version you put on the play server DEVELOPMENT.2020-07-19T21-00-50Z seems to work on my initial testing @harshavardhana

Okay cool @dpgarrick - that looks correct.

**Comment by dpgarrick** [comment]
> > I'm running Ubuntu 18.04. Don't have time currently to build a local binary on go1.13.13 but following along in slack and this version you put on the play server DEVELOPMENT.2020-07-19T21-00-50Z seems to work on my initial testing @harshavardhana
> 
> Okay cool @dpgarrick - that looks correct.

@harshavardhana  I tested that same binary some more. Uploads and downloads via mc command seem to work OK. Download from preSignedURL of >1GB file seems to also work.

However during upload of a >1GB file uploading at ~10MB/sec to a presigned URL via PUT I get a sudden failure with "Request timed out" about ~30 seconds in.

**Comment by harshavardhana** [comment]
> However during upload of a >1GB file uploading at ~10MB/sec to a presigned URL via PUT I get a sudden failure with "Request timed out" about ~30 seconds in.

That's fine @dpgarrick I don't think that is an issue - it is perhaps your disk is not keeping up with the writes coming in - `mc` is doing multipart uploads using single PUT would slow things down a bit depending on how much bandwidth you have. 

AFAICS original issue here is addressed.

**Comment by dpgarrick** [comment]
> > However during upload of a >1GB file uploading at ~10MB/sec to a presigned URL via PUT I get a sudden failure with "Request timed out" about ~30 seconds in.
> 
> That's fine @dpgarrick I don't think that is an issue - it is perhaps your disk is not keeping up with the writes coming in - `mc` is doing multipart uploads using single PUT would slow things down a bit depending on how much bandwidth you have.
> 
> AFAICS original issue here is addressed.

Performing all the same mc cp and presigned url GET/PUT operations through a client on the LAN (still through the reverse proxy) consistently achieves >100MB/s up/down speeds to the same minio server (4 disk single node) so it is not a disk issue.

When I try on external WAN clients is when I start to see this occur. I have a web front end which allows users to upload files via the presigned URL direct to minio. Using the latest release and the dev binary the uploads are failing due to timeout. I've tried on two separately located clients and it occurs for both and can confirm this does not occur when using the RELEASE.2020-06-22T03-12-50Z binary. 

I can open a new issue for this if you would like @harshavardhana

**Comment by harshavardhana** [comment]
@dpgarrick can you try 

```
mc ls play/testbucket/minio-go1.13.12.gz
[2020-07-19 14:45:48 PDT]   17MiB minio-go1.13.12.gz
```

**Comment by dpgarrick** [comment]
> @dpgarrick can you try
> 
> ```
> mc ls play/testbucket/minio-go1.13.12.gz
> [2020-07-19 14:45:48 PDT]   17MiB minio-go1.13.12.gz
> ```

Same as dev, upload stops with error request timed out. 

Seems to consistently time out at exactly 30 seconds upload elapsed regardless of upload filesize or connection speed. If upload requires less than 30 seconds it is successful.

**Comment by harshavardhana** [comment]
> Seems to consistently time out at exactly 30 seconds upload elapsed regardless of upload filesize or connection speed. If upload requires less than 30 seconds it is successful.

Yeah but there are no 30-sec timeouts anywhere :-) are you sure this is not a client timeout?

**Comment by dpgarrick** [comment]
> > Seems to consistently time out at exactly 30 seconds upload elapsed regardless of upload filesize or connection speed. If upload requires less than 30 seconds it is successful.
> 
> Yeah but there are no 30-sec timeouts anywhere :-) are you sure this is not a client timeout?

Pretty sure and this does not occur on the older release. 

There is a 30 second keep alive period in your PR cmd/http/listener.go line 86 which seems coincidental...
```
		tcpConn.SetKeepAlivePeriod(30 * time.Second)
```
@harshavardhana Can you increase this to say 45 seconds and put a binary up for me to try?

**Comment by harshavardhana** [comment]
> ```
> 		tcpConn.SetKeepAlivePeriod(30 * time.Second)
> ```

That is not a timeout @dpgarrick its keep-alive period :-)

**Comment by harshavardhana** [comment]
That value is same as it was in https://github.com/minio/minio/blob/RELEASE.2020-06-22T03-12-50Z/cmd/http/listener.go#L88

**Comment by harshavardhana** [comment]
https://github.com/minio/minio/blob/43dd68a58acdbf12f4809665432d0540af0cb123/cmd/http/server.go#L42

**Comment by harshavardhana** [comment]
AFAICS something has changed on your end @dpgarrick a timeout is different and perhaps triggered somewhere else. Just to quell even this situation we can even remove KeepAlive period

**Comment by harshavardhana** [comment]
FWIW here is the one play/testbucket/no-timeout-minio.gz without KeepAliveTimeout - so literally no timeout whatsoever anymore.

**Comment by dpgarrick** [comment]
> That is not a timeout @dpgarrick its keep-alive period :-)
> https://github.com/minio/minio/blob/43dd68a58acdbf12f4809665432d0540af0cb123/cmd/http/server.go#L42

Weird that this code calls it KeepAliveTimeout then ;) 

> FWIW here is the one play/testbucket/no-timeout-minio.gz without KeepAliveTimeout - so literally no timeout whatsoever anymore.

We did make some other changes on our upload forms to get around CORS issues we were having with minio which were fixed in one of the recent minio releases so maybe one of our changes caused this somehow.

I just tried that new binary and it is still occurring so you are correct, something on our side must be the issue here, my apologies. 

Thanks for the help on this, much appreciated.
